internal/transports/p2p/server.go

```diff
@@ -29,99 +40,259 @@ func NewServer(
+	return nil
 }
 
-func (s *server) listenAndConnect() error {
-	s.log.Info().Msgf("listening for inbound connections on port %s", s.chainParams.DefaultPort)
+func (s *server) connectToAddr(addr net.IP, port uint16) error {
+	netAddr := &net.TCPAddr{
+		IP:   addr,
+		Port: int(port),
+	}
 
-	ourAddr := net.JoinHostPort("", s.chainParams.DefaultPort)
-	listener, err := net.Listen("tcp", ourAddr)
+	conn, err := net.Dial(netAddr.Network(), netAddr.String())
 	if err != nil {
-		s.log.Error().Msgf("error creating listener, reason: %v", err)
+		s.log.Error().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("error connecting with peer, reason: %v", err)
+
+		s.log.Info().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	inbound := true
-	conn, err := listener.Accept()
+	peer, err := s.connectPeer(conn, false)
 	if err != nil {
-		s.log.Error().Msgf("error accepting connection, reason: %v", err)
+		s.log.Info().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	return s.connectPeer(conn, inbound)
+	_ = s.outboundPeers.AddPeer(peer) // don't need to check error here
+	s.addresses.UpsertPeerAddr(peer)
+	return nil
 }
 
-func (s *server) connectPeer(conn net.Conn, inbound bool) error {
-	peer, err := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log)
+func (s *server) connectPeer(conn net.Conn, inbound bool) (*peer.Peer, error) {
+	peer := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log, s)
+
+	err := peer.Connect()
 	if err != nil {
-		return err
+		peer.Disconnect()
+		s.log.Error().Str("peer", peer.String()).
+			Bool("inbound", inbound).
+			Msgf("error connecting with peer, reason: %v", err)
+		return nil, err
+	}
+
+	peer.SendGetAddrInfo()
+
+	if !inbound {
+		err = peer.StartHeadersSync()
+		if err != nil {
+			peer.Disconnect()
+			return nil, err
+		}
 	}
 
-	s.peers = append(s.peers, peer)
+	return peer, nil
+}
+
+func (s *server) observeOutboundPeers() {
+	const sleepMinutes = 5
+	time.Sleep(sleepMinutes * time.Minute)
+
+	for {
+		select {
+		case <-s.ctx.Done(): // exit if context was cancaled
+			s.log.Info().Msg("[observeOutboundPeers] exit")
+			return
+
+		default:
+			peersToConnect := s.outboundPeers.Space()
+			if peersToConnect == 0 {
+				s.log.Debug().Msg("[observeOutboundPeers] nothing to do")
+				time.Sleep(sleepMinutes * time.Minute)
+				continue
+			}
+
+			s.log.Info().Msgf("try connect with %d new peers", peersToConnect)
+
+			for ; peersToConnect > 0; peersToConnect-- {
+				s.connectToRandomAddr()
+			}
+		}
+	}
+}
+
+func (s *server) connectToRandomAddr() {
+	const tries = 20
```

Comment:
Maybe follow the same pattern as in Webhooks config, and add "MaxTries" to P2PConfig? 🤔

Reply:
It shouldn't be configurable because it's an implementation detail. 

In the end, we are attempting to establish and maintain a fixed number of connections in an infinite loop, so the value of this constant is not crucial. I use it simply to provide information to us (developers and administrators) that, over time, the implementation of the address book and its random address selection may no longer be sufficient.

Reply:
The 'tries' was removed. The AddressBook implementation was changed, and it is no longer required.